components/skin/ui/src/main/resources/PhenoTips/CreatePatientRecordForm.xml

```diff
@@ -50,28 +50,40 @@ $xwiki.jsx.use('PhenoTips.Widgets')##
     #if ($xwiki.exists('PhenoTips.StudyBindingClass'))
       ## Is there more than one sheet option?
       ## Check out all the groups this user is part of
-      #set ($toProcess = $collectionstool.queue)
-      #set ($discard = $toProcess.add($xcontext.userReference.toString()))
-      #set ($groups = [])
-      #set ($studies = [])
-      #foreach ($iteration in [0..1000])
-        #if ($toProcess.empty)
-          #break
+      #set($studiesVisibilityPreference=$xwiki.getDocument('PhenoTips.StudyClass').getObject('PhenoTips.StudyVisibilityClass').getProperty('visibility').value)
+      #if ($studiesVisibilityPreference == 'open')
+        #set ($studyClassName = 'PhenoTips.StudyClass')
+        #set ($studyTemplateName = 'PhenoTips.StudyTemplate')
+        #set ($query = $services.query.xwql("from doc.object(${studyClassName}) as std where doc.fullName &lt;&gt; :template").bindValue('template', $studyTemplateName))
+        #set ($studyDocs = $xwiki.wrapDocs($query.execute()))
+        #set ($studies = $collectionstool.sortedSet)
+        #foreach ($study in $studyDocs)
+          #set ($discard = $studies.add("$study"))
         #end
-        #set ($item = $toProcess.poll())
-        #foreach ($group in $xwiki.wrapDocs($xwiki.rightsmanager.getAllGroupsNamesForMember($item)))
-          #set ($groupObj = $group.getObject('PhenoTips.PhenoTipsGroupClass'))
-          #if (!$groupObj &amp;&amp; $group.documentReference.name.endsWith(' Administrators'))
-            #set ($groupObj = $xwiki.getDocument($services.model.resolveDocument($stringtool.removeEnd($group.documentReference.name, ' Administrators'), $group.documentReference.parent)).getObject('PhenoTips.PhenoTipsGroupClass'))
+      #else
+        #set ($toProcess = $collectionstool.queue)
+        #set ($discard = $toProcess.add($xcontext.userReference.toString()))
+        #set ($groups = [])
+        #set ($studies = [])
```

Comment:
This code here only checks if there are any studies that apply to the current user, so it doesn't care about the order. For performance, this could be greatly simplified: the `open` case could just `count()` the number of studies and check that it's `> 0`, while the `restricted` case could `#break` at the first study found.